src/test/ui/pathless-extern-ok.rs

```diff
@@ -0,0 +1,9 @@
+// edition:2018
+// compile-flags:--extern alloc
```

Comment:
Is this achieved by whitelist?

Reply:
In a sense, yes.  The crate must be marked with `#![stable]`, otherwise the `-Zforce-unstable-if-unmarked` makes it unstable.

Reply:
Ah... so just for my education, what happens if `-Z force-unstable-if-unmarked` isn't there? (And it cannot be there on stable when you are not using cargo since it's a `-Z` flag?)

Reply:
It will successfully link whatever library is in the given path.  That library will be added to the extern prelude with the identifier "alloc".

Reply:
Isn't that a problem then if you are not using cargo since it would allow you to use unstable stuff on stable? I'm rather surprised that `-Z force-unstable-if-unmarked` isn't the default and that you need a flag to *opt-out* for sysroot crates.

Reply:
Sorry, I'm not understanding the question.

On stable `rustc`, you cannot access any crates that are marked "unstable".  Either from `--extern foo` or `--extern foo=path/to/sysroot/libfoo.rlib` or `extern crate foo;` or `-L path/to/unstable/crates`.  

On nightly, you can include the attribute `#![feature(rustc_private)]` to override this check.

All of the sysroot crates are built with the [`-Zforce-unstable-if-unmarked`](https://github.com/rust-lang/rust/blob/master/src/bootstrap/bin/rustc.rs#L149-L155) flag.  

The opt-in nature is just saying `#![stable]` crates are OK to access, even with the `-Zforce-unstable-if-unmarked` flag (that is the "if unmarked" part).  I believe the only crates marked stable are core, std, alloc, and proc_macro. 

In theory, anyone can compile any crate using unstable features using nightly, and then load those crates from stable.  In practice, you can't (due to rustc version checking) or clearly something you shouldn't do (like set `RUSTC_BOOTSTRAP` env var).

Can you say more how you think that can be circumvented?

Reply:
> All of the sysroot crates are built with the [`-Zforce-unstable-if-unmarked`](https://github.com/rust-lang/rust/blob/master/src/bootstrap/bin/rustc.rs#L149-L155) flag.

Oh! -- I missed this part. I thought you meant that when the sysroot crate is used, unless there is a `-Zforce...`, then it won't be gated... but I see now that this is about when the crate is built, not used. Now it clicks. Maybe this conversation should be distilled into the rustc guide?

Reply:
Done: https://github.com/rust-lang/rustc-guide/pull/462